nomad/structs/diff.go

```diff
@@ -606,17 +611,48 @@ func serviceCheckDiff(old, new *ServiceCheck, contextual bool) *ObjectDiff {
 // occurred.
 func checkHeaderDiff(old, new map[string][]string, contextual bool) *ObjectDiff {
 	diff := &ObjectDiff{Type: DiffTypeNone, Name: "Header"}
+	var oldFlat, newFlat map[string]string
+
 	if reflect.DeepEqual(old, new) {
 		return nil
 	} else if len(old) == 0 {
 		diff.Type = DiffTypeAdded
+		newFlat = flatmap.Flatten(new, nil, false)
 	} else if len(new) == 0 {
 		diff.Type = DiffTypeDeleted
+		oldFlat = flatmap.Flatten(old, nil, false)
+	} else {
+		diff.Type = DiffTypeEdited
+		oldFlat = flatmap.Flatten(old, nil, false)
+		newFlat = flatmap.Flatten(new, nil, false)
+	}
+
+	diff.Fields = fieldDiffs(oldFlat, newFlat, contextual)
+	return diff
+}
+
+// checkRestartDiff returns the diff of two service check check_restart
+// objects. If contextual diff is enabled, all fields will be returned, even if
+// no diff occurred.
+func checkRestartDiff(old, new *CheckRestart, contextual bool) *ObjectDiff {
+	diff := &ObjectDiff{Type: DiffTypeNone, Name: "CheckRestart"}
+	var oldFlat, newFlat map[string]string
```

Comment:
The problem arose from how the maps were being constructed. See the commit I pushed to see the change. Can you add a test for headers as well since I think it is likely they were misbehaving as well.

Reply:
Interesting! Will do! Thanks!